[Client] Use async rpc for remote call and actor creation

mwtian opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
Currently in Ray client, each call to schedule a task, an actor or an actor method requires the client to wait synchronously for object or actor ID(s) to return from the cluster. Running the following expression has been slow when there is non-negligible latency between Ray client and cluster: `refs = [f.remote() for _ in range(1000)]`. Each `f.remote()` call blocks on a client-cluster round trip.

This change allows `ClientObjectRef` and `ClientActorRef` to be created with a `Future` instead of a `bytes` ID, making the RPC response from `Schedule` calls non-blocking. Underlying object and actor IDs need only to be waited for when these references are used.

Additional changes:
- Scheduling calls are moved to use streaming RPC, together with other operations like `get` and `put`. So task and method calls are still serialized.
- For name actors created from `get_actor()`, they lack metadata such as per method signature and `num_returns`. Make a separate call to retrieve `num_returns` and method signatures.
- Check method signatures at client, so callers can still get function signature mismatch errors synchronously. This behavior is the same as when running as a driver.

Performance is tested by running Ray client from a laptop, and a Ray cluster with 1 `m5.2xlarge` node in AWS `us-west-2`.  Before the change,
| Measurements | Time in seconds |
| --------------- | ---------------- |
| 40 remote actor method calls (walltime in script) | 1.437 |
| 40 remote actor method calls + get 40 ObjectRefs in a list (walltime in script) | 1.488 |
Time to start calls (from before making actor method calls, to method called) | max: 1.4865, min: 0.0829, median: 0.7895 |

After the change,
| Measurements | Time in seconds |
| --------------- | ---------------- |
| 40 remote actor method calls (walltime in script) | 0.004 |
| 40 remote actor method calls + get 40 ObjectRefs in a list (walltime in script) | 0.095 |
Time to start calls (from before making actor method calls, to method called) | max: 0.1090, min: 0.0863, median: 0.0987 |

Round trip latency between client and cluster is mostly hidden from remote calls now.

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

mwtian commented:
cc @iycheng

AmeerHajAli commented:
Wow @mwtian this looks really good. Do you have early performance numbers?

ijrsvt commented on python/ray/includes/object_ref.pxi:
```diff
@@ -88,8 +88,8 @@ cdef class ObjectRef(BaseID):
 
-    cdef size_t hash(self):
-        return self.data.Hash()
+    def size(self):
```
can you revert these changes?

ijrsvt replied:
Ahh okay!

ijrsvt commented on python/ray/includes/unique_ids.pxi:
```diff
@@ -323,10 +328,40 @@ cdef class ClientActorRef(ActorID):
+
+    cdef size_t hash(self):
+        self._wait_for_id()
+        return self.data.Hash()
+
+    cdef CActorID native(self):
+        self._wait_for_id()
+        return <CActorID>self.data
+
     @property
     def id(self):
         return self.binary()
 
+    cdef _set_id(self, id):
+        check_id(id, CActorID.Size())
+        self.data = CActorID.FromBinary(<c_string>id)
+        client.ray.call_retain(id)
+
```
Could you add some tests to ensure that the 'public API' for ClientObjects and non-client Objects are the same?

ijrsvt commented on python/ray/util/client/worker.py:
```diff
@@ -313,31 +314,39 @@ def wait(self,
         return (client_ready_object_ids, client_remaining_object_ids)
 
-    def call_remote(self, instance, *args, **kwargs) -> List[bytes]:
+    def call_remote(self, instance, *args, **kwargs) -> List[Future]:
```
Can you make sure that `call_remote` in api.py also has the updated type? I think it might also be good to change `call_remote` in api.py to `_call_remote` to make it clear that it is not for public usage.

ijrsvt replied:
A TODO is totally fine! I think moving all of those to internal would be good in a follow up :)

ijrsvt replied:
```suggestion
    def call_remote(self, instance, *args, **kwargs) -> List[Future]:
    # TODO(mwtian) Refactor `call_remote`, `call_release` & `call_retain`
    # to have internal prefix
```

ijrsvt reviewed (commented):
Looks pretty good overall! Just some initial feedback!

mwtian commented:
> Wow @mwtian this looks really good. Do you have early performance numbers?

Microbenchmarks do show a lot of difference. Trying to run it in a cluster environment. Will update when the numbers are available.

mwtian commented:
> Wow @mwtian this looks really good. Do you have early performance numbers?

Updated latency stats in the PR description. Basically most of client-cluster round trip latency is hidden from remote calls now.

ijrsvt commented on python/ray/tests/BUILD:
```diff
@@ -316,7 +316,7 @@ py_test_module_list(
     "test_list_actors.py",
   ],
-  size = "medium",
+  size = "large",
```
Did this test start taking longer because of this PR?

mwtian replied:
There seems to be an issue with test_basic_2.py test_actor_recursive, but I'm not able to reproduce it locally. Still investigating.

ijrsvt commented on python/ray/tests/test_basic_2.py:
```diff
@@ -60,9 +60,9 @@ def f2(x, y, *args):
     assert ray.get(f1.remote(1)) == (1, )
     assert ray.get(f1.remote(1, 2, 3)) == (1, 2, 3)
-    with pytest.raises(Exception):
+    with pytest.raises(TypeError):
```
Are these changes necessary for the test to work, or are they just making the tests more specific (better)

mwtian replied:
Making the tests more specific for both client and driver modes.

ijrsvt replied:
Fantastic :) This is definitely helpful!

ijrsvt commented on python/ray/util/client/worker.py:
```diff
@@ -500,14 +514,12 @@ def internal_kv_list(self, prefix: bytes) -> bytes:
         if self.server is not None:
             return self.get_cluster_info(
-                ray_client_pb2.ClusterInfoType.IS_INITIALIZED)
+                ray_client_pb2.ClusterInfoType.IS_INITIALIZED) is not None
```
I think the return value from the server is a boolean (T/F). I think we want to keep directly returning it.

mwtian replied:
Reverted.

mwtian commented:
Rebased with on master branch. Ready for reviews now.

AmeerHajAli commented on python/ray/tests/test_asyncio.py:
```diff
@@ -304,15 +304,22 @@ def interceptor(e):
     del x1
-    wait_for_condition(lambda: num_exceptions == 1)
+    if client_test_enabled():
+        # In client mode, x1 may not have received the underlying exception
+        # when deleted, so no exception is raised.
+        prev_num_exceptions = num_exceptions
```
is this a bug? Should it behave like native ray?

mwtian replied:
I think not really. Exception handling and behavior during {Client}ObjectRef destruction is not a public API. Client uses async communication with cluster, but scripts running in cluster can communicate with workers synchronously.

AmeerHajAli commented on python/ray/tests/test_asyncio.py:
```diff
@@ -304,15 +304,22 @@ def interceptor(e):
     del x1
     await asyncio.sleep(1)
-    assert num_exceptions == 1, num_exceptions
+    assert num_exceptions == prev_num_exceptions, (num_exceptions,
```
maybe explicitly write expected number of exceptions: num_exceptions received number of exceptions: prev_num_exceptions in the error message

AmeerHajAli commented on python/ray/tests/test_client.py:
```diff
@@ -135,14 +135,21 @@ def f():
+
+        def get(req, metadata=None):
+            nonlocal get_count
+            get_count += 1
+            return get_stub(req, metadata=metadata)
+
+        ray.worker.server.GetObject = get
+
         refs = [f.remote() for _ in range(100)]
-        with ray.worker.data_client.lock:
-            req_id_before = ray.worker.data_client._req_id
         assert ray.get(refs) == ["OK" for _ in range(100)]
+
         # Only 1 RPC should be sent.
-        with ray.worker.data_client.lock:
-            assert ray.worker.data_client._req_id == req_id_before + 1, \
-                ray.worker.data_client._req_id
+        assert get_count == 1
```
I don't understand this change. Can you please explain?

mwtian replied:
Basically, `ray.get()` is converted to use unary-unary RPC instead of streaming RPC, so we need another way to check if only 1 RPC request is sent. `ray.get()` does not need to serialize with other operations. In future it will be converted to unary-streaming to chunk response data.

AmeerHajAli commented on python/ray/tests/test_client.py:
```diff
@@ -331,16 +338,27 @@ def __init__(self):
 
             def say_hello(self, whom):
                 self.count += 1
-                return ("Hello " + whom, self.count)
+                return "Hello " + whom, self.count
+
+            @ray.method(num_returns=2)
+            def say_hi(self, whom):
+                self.count += 1
+                return "Hi " + whom, self.count
```
why this is needed?

mwtian replied:
Test `@ray.method(num_returns=2)` is handled correctly.

AmeerHajAli commented on python/ray/tests/test_client.py:
```diff
@@ -331,16 +338,27 @@ def __init__(self):
             def say_hello(self, whom):
                 self.count += 1
-                return ("Hello " + whom, self.count)
+                return "Hello " + whom, self.count
+
+            @ray.method(num_returns=2)
+            def say_hi(self, whom):
+                self.count += 1
+                return "Hi " + whom, self.count
 
         actor = HelloActor.remote()
         s, count = ray.get(actor.say_hello.remote("you"))
         assert s == "Hello you"
         assert count == 1
-        s, count = ray.get(actor.say_hello.remote("world"))
+
+        ref = actor.say_hello.remote("world")
+        s, count = ray.get(ref)
         assert s == "Hello world"
         assert count == 2
 
+        r1, r2 = actor.say_hi.remote("ray")
+        assert ray.get(r1) == "Hi ray"
+        assert ray.get(r2) == 3
```
Is this to improve the test? Shouldn't it behave like it was before making this change?

AmeerHajAli replied:
(just double checking to make sure the async behavior did not introduce an issue)

mwtian replied:
Yes, this is an enhancement to ensure the behavior is the same between cluster and client modes, for `@ray.method(num_returns=2)`.

AmeerHajAli commented on python/ray/tests/test_client_terminate.py:
```diff
@@ -113,11 +113,12 @@ def f():
 
-        def mock_terminate(term, metadata):
-            raise MetadataIsCorrectlyPassedException(metadata[1][0])
+        def mock_terminate(self, term):
```
what term is used for?

mwtian replied:
This mocks the `Terminate()` handler, which accepts the `TerminateRequest` (`term`). We need the same function signature.

AmeerHajAli commented on python/ray/tests/test_client_warnings.py:
```diff
@@ -19,33 +17,6 @@ def reset_debug_logs():
 class LoggerSuite(unittest.TestCase):
     """Test client warnings are raised when many tasks are scheduled"""
 
-    def testManyTasksWarning(self):
-        with ray_start_client_server() as ray:
-
-            @ray.remote
-            def f():
-                return 42
-
-            with self.assertWarns(UserWarning) as cm:
-                for _ in range(TASK_WARNING_THRESHOLD + 1):
-                    f.remote()
-            assert f"More than {TASK_WARNING_THRESHOLD} remote tasks have " \
-                "been scheduled." in cm.warning.args[0]
-
-    def testNoWarning(self):
-        with ray_start_client_server() as ray:
-
-            @ray.remote
-            def f():
-                return 42
-
-            with warnings.catch_warnings(record=True) as warn_list:
-                for _ in range(TASK_WARNING_THRESHOLD):
-                    f.remote()
-            assert not any(f"More than {TASK_WARNING_THRESHOLD} remote tasks "
-                           "have been scheduled." in str(w.args[0])
-                           for w in warn_list)
-
```
Why are we removing this? this is one of our important usability warnings.

AmeerHajAli replied:
CC @ckw017

mwtian replied:
Yes, because it is no longer necessary with async task calls.

AmeerHajAli replied:
Wouldn't the overall performance be the same if the network BW is abused?

mwtian replied:
Each task is very small e.g. < 1kb. The issue has been blocking on round trip latency instead of throughput, which is resolved now. And in general, having unnecessary warnings can confuse users and degrade the experience.

AmeerHajAli replied:
Should we also remove the logic this is testing?

ckw017 replied:
Sounds good to me, avoiding the round-trip cost should definitely be a big help for this case.

pcmoritz commented:
There is some complexity introduced here since the client now has to know both the future AND the object reference (the latter will become available later). Is it necessary for the client to know the object reference? I wonder if the cleaner design would be to have only a client side ID/future for every .remote call on the client side and always resolve that to the underlying object reference on the server side. This will get rid of the awkward mutexes introduced in this PR.

pcmoritz commented on python/ray/tests/test_client.py:
```diff
@@ -454,8 +472,11 @@ def print_on_stderr_and_stdout(s):
+                ValueError, match="Failed to look up actor with name 'abc'"):
+            a = ray.get_actor("abc")
+            # Reading underlying ActorID to raise the Exception from server.
+            print(a.id)
```
This really shouldn't be needed, please make the test pass without this!

mwtian replied:
Merge issue that left an obsolete logic. Will be removed!

mwtian commented:
> There is some complexity introduced here since the client now has to know both the future AND the object reference (the latter will become available later). Is it necessary for the client to know the object reference? I wonder if the cleaner design would be to have only a client side ID/future for every .remote call on the client side and always resolve that to the underlying object reference on the server side. This will get rid of the awkward mutexes introduced in this PR.

I thought about this approach at the beginning. One complexity / inefficiency is that if the user uses the ClientObjectRef / ClientActorRef as task args or captures the references in tasks / actors, we need to run another lookup to find the corresponding ref in the cluster. Another the downside is that if some users managed to use ClientObjectRef / ClientActorRef out-of-band, i.e. using raw ids, switching to client-side ref would break them too.

pcmoritz commented on python/ray/tests/test_client_references.py:
```diff
@@ -261,7 +285,8 @@ def check(self):
             api = _ClientContext()
             api.connect("localhost:50051", namespace="default_test_namespace")
-            api.get_actor("actor")
+            # Make a call to resolve the underlying ActorID.
```
Again this shouldn't be needed

mwtian replied:
Will revert.

pcmoritz commented on python/ray/util/client/common.py:
```diff
@@ -217,61 +247,89 @@ def __init__(self,
-                dict(
-                    inspect.getmembers(actor_class.actor_cls,
-                                       is_function_or_method)).keys())
+            self._method_num_returns = {}
```
I suggest splitting this PR into two: One for the async submission, and one (smaller one) for the num_returns fix and the `data = ray.worker._dumps_from_client(self.actor_cls)` change above.

Maybe you have ideas how it could be split even more.

It is quite possible that this PR will cause problems further down the line (just by the nature of the change), so having it chunked into smaller pieces will make it much easier to debug problems.

mwtian replied:
Good suggestion, since we need to wait for other changes to merge too. I'm thinking of a preliminary change that allows 
1.  Client{Object,Actor}Ref to take futures
2. Store num_returns & function signatures, check them synchronously
(2) is needed by the async change for the (desirable) behavior of raise `TypeError` synchronously when functions are called with incorrect arguments. They should have no / minimal conflict with other PRs.

mwtian replied:
Splitted out a (majority) part of the code to #18677, which should not have significant behavior change.

ericl commented:
Is this really a release blocker? It looks more like a performance optimization than a bug.

mwtian commented:
> Is this really a release blocker? It looks more like a performance optimization than a bug.

It seems to be a significant usability issue frequently complained by Ray client users, although there are workarounds. It is ok to not block the release on this, if it does not meet the bar.

AmeerHajAli commented on python/ray/tests/test_placement_group_2.py:
```diff
@@ -46,28 +46,18 @@ def value(self):
-        try:
+        with pytest.raises(ValueError, match="bundle index -2 is invalid"):
             Actor.options(
                 placement_group=placement_group,
                 placement_group_bundle_index=-2).remote()
-        except ValueError:
-            error_count = error_count + 1
-        assert error_count == 2
 
-        try:
+        with pytest.raises(ValueError, match="bundle index must be -1"):
             Actor.options(placement_group_bundle_index=0).remote()
-        except ValueError:
-            error_count = error_count + 1
-        assert error_count == 3
 
```
is this some unrelated fixes/improvements to the test?

mwtian replied:
Yes, this is the make the test more debuggable.

ijrsvt reviewed (commented):
I'm a bit concerned with the test changes. What happens without them?

mwtian commented on python/ray/includes/object_ref.pxi:
```diff
@@ -166,8 +166,19 @@ cdef class ClientObjectRef(ObjectRef):
+            # instead.
+            except Exception:
+                logger.info(
+                    "Exception in ObjectRef is ignored in destructor. "
+                    "To raise the exception, use the object reference before "
+                    "its destructor runs.")
```
Exceptions cannot be raised from destructors and trying to print them can crash too. If the exception is a concern to users, it can be acted upon by calling `ray.get()` on the object ref before the object ref is destroyed, to raise the exception. Is there a better way to reword this?

ijrsvt replied:
Let me make take a stab at fixing it!

iycheng commented on python/ray/util/client/worker.py:
```diff
@@ -138,9 +136,7 @@ def __init__(
-        # Track these values to raise a warning if many tasks are being
-        # scheduled
-        self.total_num_tasks_scheduled = 0
+        # Track these values to raise a warning if too data are transferred.
```
# Track these values to raise a warning if a lot of data is transferred.

iycheng replied:
# Track these values to raise a warning if a lot of data is transferred.

iycheng commented on python/ray/includes/unique_ids.pxi:
```diff
@@ -326,8 +329,19 @@ cdef class ClientActorRef(ActorID):
-            client.ray.call_release(self.id)
+        if client.ray.is_connected():
+            try:
+                self._wait_for_id(timeout=10)
```
The question here: any reason you pick 10s? Another question is that if it failed to be resolved, will it lead to a leak?

mwtian replied:
The timeout shouldn't be too long because it is in destructor, and the user doesn't need the ref / object anymore. We can lower the timeout to 5 or 2. Timeout for `ray.get()` is 2s.

When there is a failure to resolve an ID, the ID will not be released so the ref will be kept server-side after the client ref is gone. Object and actor refs are still cleaned up server-side, when client-server shuts down. Failure to resolve an ID most likely happens after a failed connection or server. So this should not be a significant issue. Before this change, this issue can happen in theory if a ClientTaskTicket response is lost, although this is more rare.

iycheng replied:
I feel this is somewhat still important. It's still a leak and under unstable internet environment it's going to happen a lot. Although it will be released once the client exits, it still leaks in the code and that program will fail.

Btw, how is the reconnect going to work here? Will the failed resolving be called again?

And is it better to retry this and fail the program in the extreme case (for example failed to do cleanup for 5 times)

Or should we make the timeout here to be infinite?

mwtian replied:
When the resolving takes too long and fails, if the resolution succeeds in the background with a server response, the ID future will be set but there will be not ref counting. I thought about it for a bit, and it seems there is no issue with using an infinite timeout here. It becomes closer to the behavior before this PR, where client waits synchronously for response from server.

iycheng commented on src/ray/protobuf/ray_client.proto:
```diff
@@ -367,6 +367,9 @@ message DataRequest {
     AcknowledgeRequest acknowledge = 9;
+    ClientTask task = 10;
+    TerminateRequest terminate = 11;
+    ClientListNamedActorsRequest list_named_actors = 12;
```
It looks like we move some functions to the data client. Should we clean up the raylet one, like Schedule/ListNamedActors ?

mwtian replied:
It is a bit confusing but the handlers of Schedule / ListNamedActors in server.py are called by the streaming Datapath handler to handle the corresponding logic e.g. ClientTask / ClientTaskTicket. Before this change, Get / Put etc are in the same situation, where client - server interacts via Datapath but Get / Put handlers are still exposed. This can be cleaned up in a future PR, after we have a more stable client / server protocol.

iycheng replied:
Got it. It makes sense!

iycheng reviewed (commented):
Overall it looks clean, I only have one big concern about the leak.

Another thing is that should we clean up the old code? It looks like we've moved some functions to the dataclient.

iycheng commented:
> > @mwtian , can we please add more unit tests to the different stubs and in general?
> > Can we also please add the benchmarks or this is already in the existing benchmarks and we should just make sure it does not regress somehow?
> 
> 1. There are existing unit tests for `ray.remote(...)`, `ray.kill(...)` and `ray.list_named_actors(...)`. They still apply here. Testing the RPC stubs directly seems like testing the internal implementations, which Ray is not currently doing. Maybe they are useful in edge cases, but that can be added in future.
> 2. Existing unit tests are mostly CPU bound. The `client: tasks and get batch` benchmark exercises this code path pretty heavily, but don't show a lot of difference. We don't yet have tests with simulated RPC latencies, but we probably should add them in future.
>    Before the change, `client: tasks and get batch per second 1.0 +- 0.0`
>    After the change, `client: tasks and get batch per second 1.03 +- 0.0`
> 
> The result from manually tested more realistic scenario is posted at the beginning.

I feel the microbenchmark is not very useful for some cases. Do you want to add a new one for this work?

mwtian commented:
> I feel the microbenchmark is not very useful for some cases. Do you want to add a new one for this work?

This work mostly hides round trip latencies between client and server for scheduling. But within a single machine there is not much latency. I can try injecting latencies but it could be a larger change. Maybe in another PR?

iycheng commented:
> > I feel the microbenchmark is not very useful for some cases. Do you want to add a new one for this work?
> 
> 
> 
> This work mostly hides round trip latencies between client and server for scheduling. But within a single machine there is not much latency. I can try injecting latencies but it could be a larger change. Maybe in another PR?

Thanks for the explanation! I'll treat it as the limitation of micro benchmark. Let's don't do that. We should add a metric later in nightly test for this.

AmeerHajAli reviewed (approved):
@mwtian , Thanks for following through with this. It looks very close and people seem to be excited about this. Let me know when you finish addressing the comments and run some manual testing in the product, perhaps @ckw017 can help with the testing (or kill the proxy).

ijrsvt commented on python/ray/includes/object_ref.pxi:
```diff
@@ -166,8 +166,19 @@ cdef class ClientObjectRef(ObjectRef):
+            except Exception:
+                logger.info(
+                    "Exception in ObjectRef is ignored in destructor. "
+                    "To raise the exception, use the object reference before "
```
Maybe something like:
```
To receive this exception in application code, ray.get() the object 
reference before its destructor is run.
```

mwtian replied:
Thanks for the suggestion!

iycheng reviewed (approved):
Nice job!

mwtian commented:
> @mwtian , Thanks for following through with this. It looks very close and people seem to be excited about this. Let me know when you finish addressing the comments and run some manual testing in the product, perhaps @ckw017 can help with the testing (or kill the proxy).

I believe all comments have been addressed. The change has been tested in product with this [wheel](https://ray-wheels.s3.us-west-2.amazonaws.com/test_wheels/mwtian-local-ref/518e5601d9a3b4e5b960bc90246afda38b56ef37/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl). Reconnection after network disruption works. Also the following code snippet took ~2s with the change instead of ~100s,
```
start = time.perf_counter()
result = [f.remote(i) for i in range(1000)]
ray.get(result)
end = time.perf_counter()
```
This should be ready to be merged now.

AmeerHajAli commented:
Thanks @mwtian !!!!
